### DISCUSSION TITLE
A problem about the body force

### DISCUSSION THREAD

**Discussion by Johnney12138** [idaholab/moose#23784] [Q&A Modules: Navier-Stokes] [answered]
Hello,
I was trying to solve the 2D Navier-Stokes equation. I want to add the body force to the momentum equation, and the expression of the body force is Vy*Cos(x)*Sin(y)-Vx*Sin(y)^2 in the x-axis, Vx*Cos(x)*Sin(y)-Vy*Cos(x)^2 in the y-axis, where Vx and Vy are velocity components. I tried many combinations of codes, but the programs always had bugs. The reason for the bugs is the BodyForce block needs to add the expressions from the Function block. How to solve this problem?

Thanks.
The complete code is attached as follows:
`[Mesh]
    [gen]
      type = GeneratedMeshGenerator
      dim = 2
      xmin = 0
      xmax = 1.0
      ymin = 0
      ymax = 1.0
      nx = 64
      ny = 64
      elem_type = QUAD9
    []
    [./corner_node]
      type = ExtraNodesetGenerator
      new_boundary = 'pinned_node'
      nodes = '0'
      input = gen
    [../]
  []
  
  [AuxVariables]
    [./vel_x]
      order = SECOND
    [../]
    [./vel_y]
      order = SECOND
    [../]
  []
  
  [AuxKernels]
    [./vel_x]
      type = VectorVariableComponentAux
      variable = vel_x
      vector_variable = velocity
      component = 'x'
    [../]
    [./vel_y]
      type = VectorVariableComponentAux
      variable = vel_y
      vector_variable = velocity
      component = 'y'
    [../] 
    [./bodyfx]
      type = FunctionAux
      variable = bodyfx
      function = 'vel_y*cos(x)*sin(y) - vel_x*sin(y)*sin(y)'
    [../]  
    [./bodyfy]
      type = FunctionAux
      variable = bodyfy
      function = 'vel_x*cos(x)*sin(y) - vel_y*cos(x)*cos(x)'
    [../]   
  []
  
  [Variables]
    [./velocity]
      order = SECOND
      family = LAGRANGE_VEC
    [../]
  
    [./T]
      order = SECOND
      [./InitialCondition]
        type = ConstantIC
        value = 1.0
      [../]
    [../]
  
    [./p]
    [../]
  []
  
  [Kernels]
    [./mass]
      type = INSADMass
      variable = p
    [../]
  
    [./momentum_time]
      type = INSADMomentumTimeDerivative
      variable = velocity
    [../]
  
    [./momentum_convection]
      type = INSADMomentumAdvection
      variable = velocity
    [../]
  
    [./momentum_viscous]
      type = INSADMomentumViscous
      variable = velocity
    [../]
  
    [./momentum_pressure]
      type = INSADMomentumPressure
      variable = velocity
      pressure = p
      integrate_p_by_parts = true
    [../]
      
    [./gravity]
      type = INSADGravityForce
      variable = velocity
      gravity = '0 -9.81 0'
    [../]

    [./bodyforce]      
      type = VectorBodyForce
      variable = velocity
      function_x = 'v_y*cos(x)*sin(y) - v_x*sin(y)*sin(y)'
      function_y = 'v_x*cos(x)*sin(y) - v_y*cos(x)*cos(x)'
    [../]
  
   [./temperature_time]
     type = INSADHeatConductionTimeDerivative
     variable = T
   [../]
  
   [./temperature_advection]
     type = INSADEnergyAdvection
     variable = T
   [../]
  
   [./temperature_conduction]
     type = ADHeatConduction
     variable = T
     thermal_conductivity = 'k'
   [../]
  []
  
  [BCs]
    [./no_slip]
      type = VectorFunctionDirichletBC
      variable = velocity
      boundary = 'bottom right left'
    [../]
  
    [./lid]
      type = VectorFunctionDirichletBC
      variable = velocity
      boundary = 'top'
      function_x = 'lid_function'
    [../]
  
    [./T_hot]
      type = DirichletBC
      variable = T
      boundary = 'top'
      value = 1
    [../]
  
    [./T_cold]
      type = DirichletBC
      variable = T
      boundary = 'bottom'
      value = 0
    [../]
  
    [./pressure_pin]
      type = DirichletBC
      variable = p
      boundary = 'pinned_node'
      value = 0
    [../]
  []
  
  [Materials]
    [./const]
      type = ADGenericConstantMaterial
      prop_names = 'rho mu cp k'
      prop_values = '8  1.5  0.4  50'
    [../]
    [./ins_mat]
      type = INSAD3Eqn
      velocity = velocity
      pressure = p
      temperature = T
    [../]  
    [./v_x]
      type = ParsedMaterial
      f_name = 'v_x'
      function = 'vel_x'
      args = 'vel_x'
    [../]
    [./v_y]
      type = ParsedMaterial
      f_name = 'v_y'
      function = 'vel_y'
      args = 'vel_y'
    [../]
  []
  
  [Functions]
    [./lid_function]
      type = ParsedFunction
      expression = '2*x-3*y'
    [../]
  []
  
  [Preconditioning]
    [./SMP]
      type = SMP
      full = true
      solve_type = 'NEWTON'
    [../]
  []
  
  [Executioner]
    type = Transient
    num_steps = 400
    dt = .0005
    dtmax = .005
    petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels'
    petsc_options_value = 'asm      2               ilu          4'
    line_search = 'none'
    nl_rel_tol = 1e-5
    nl_abs_tol = 1e-6
    nl_max_its = 6
    l_tol = 1e-6
    l_max_its = 500
  []
  
  [Outputs]
    file_base = lid_driven_out
    exodus = true
    perf_graph = true
  []`

**Comment by GiudGiud** [comment, marked as answer]
Hello

You cannot pass the whole function to the BodyForce kernels.
You need to define them in the Functions block as ParsedFunction objects, then pass that to the BodyForce kernel

Guillaume

**Reply by Johnney12138** [reply]
Hi @GiudGiud ,
Thank you for your reply.
According to your suggestions, I have defined the functions in the Functions block as ParsedFunction object. This creates a new problem: the functions include the velocity components Vx and Vy, but the velocity components created in AuxVariables are not identified by the ParsedFunction object.

**Reply by GiudGiud** [reply]
Functions do not support dependencies on variables unfortunately. 

You will need to use a material property instead. Such as this one
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html

then use a MatBodyForce kernel (we dont have it, you ll need to create it)
You could try using https://mooseframework.inl.gov/source/kernels/ADMatCoupledForce.html but with a coupled variable of 1 (so the term becomes 1 * material_property) but not sure it will work

**Reply by Johnney12138** [reply]
Thank you very much for your reply. I will try to solve this problem according to your suggestions.

**Comment by oh304** [comment]
Hi Johnney, did you fix this? I'm having a similar issue and can't get it right. If you have could I please see your working code?
Best wishes,

Oscar

**Reply by Johnney12138** [reply]
Hi, Oscar. I haven't coded yet. 
Best,
Johnney
